recirq/contextuality/magic_square_game.py

```diff
@@ -0,0 +1,749 @@
+        bob_choices = np.zeros((3, 3, repetitions, 3), dtype=bool)
+
+        alice_choices[0, :, :, 0] = self.alice_measurements[0, :, :, 1]
+        alice_choices[0, :, :, 1] = self.alice_measurements[0, :, :, 0]
+        alice_choices[1, :, :, :2] = self.alice_measurements[1, :, :, :2]
+        alice_choices[2, :, :, :2] = 1 - self.alice_measurements[2, :, :, :2]
+        bob_choices[:, 0, :, 0] = self.bob_measurements[:, 0, :, 1]
+        bob_choices[:, 0, :, 1] = self.bob_measurements[:, 0, :, 0]
+        bob_choices[:, 1:, :, :2] = self.bob_measurements[:, 1:, :, :2]
+
+        alice_choices[:, :, :, 2] = self.alice_measurements[:, :, :, 2]
+        bob_choices[:, 0, :, 2] = 1 - self.bob_measurements[:, 0, :, 2]
+        bob_choices[:, 1, :, 2] = 1 - self.bob_measurements[:, 1, :, 2]
+        bob_choices[:, 2, :, 2] = self.bob_measurements[:, 2, :, 2]
+
+        return alice_choices, bob_choices
+
+    def generate_choices(
+        self,
+        game: Literal[
+            "guess_3rd",
+            "measure_3rd_classical_multiplication",
+            "measure_3rd_quantum_multiplication",
+        ],
+        seed: int | None = None,
+    ) -> tuple[np.ndarray, np.ndarray]:
+        """Generate choices from Alice and Bob's measurements.
+
+        Args:
+            seed: A seed for the random number generator.
+            game:
+                guess_3rd means making two measurements and infering the third bit
+                measure_3rd_classical_multiplication: make 4 measurements for Alice and 4 for Bob.
+                get three bits out of them by mutiplying two of them toghether.
+                This corresponds to measure A and B to compute A*B.
+                measure_3rd_quantum_multiplication: make 3 measurements for Alice and 3 for Bob.
+                Use a two qubit interaction to directly measure A*B.
+
+        Returns:
+            Alice and Bob's choices in the game.
+
+        Raises:
+            NotImplementedError: If Alice and Bob measure unequal numbers of Paulis.
+        """
+        # return self._select_choices_from_second_data()
+        if game == "guess_3rd":
+            return self._generate_choices_from_rules_guess_3rd()
+        if game == "measure_3rd_classical_multiplication":
+            return self._generate_choices_from_rules_measure_3rd_classical_multiplication()
+        if game == "measure_3rd_quantum_multiplication":
+            return self._generate_choices_from_rules_measure_3rd_quantum_multiplication()
+
+    def get_win_matrix(self, game, seed: int | None = None) -> np.ndarray:
+        """Find the fraction of the time that
+        Alice and Bob "agree" (in the intersection)
+        given that
+        they "multiply correctly" (alice to -1 and bob to +1).
+
+        Args:
+            seed: The seed for the random number generator.
+            game:
+                guess_3rd means making two measurements and infering the third bit
```

Comment:
I would probably put quotes around "guess_3rd" and other literals.  Same with identical docstrings below.